src/flb_pack.c

```diff
@@ -1512,9 +1512,31 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
                 } else if (v->type == MSGPACK_OBJECT_STR){
                     val     = v->via.str.ptr;
                     val_len = v->via.str.size;
-                    if ( val_len != 1 || val[0] < '0' || val[0] > '7' ) {
+                    if ( !(val_len == 1 && val[0] >= '0' && val[0] <= '7' )) {
+                        char* allowed_levels[] = {
+                            "emerg",
+                            "alert",
+                            "crit",
+                            "err",
+                            "warning",
+                            "notice",
+                            "info",
+                            "debug"
+                        };
+                        const size_t array_size = sizeof(allowed_levels) / sizeof(allowed_levels[0]);
+                        int i;
+                        
+                        for (i = 0; i < array_size; ++i) {
+                            if (!strncmp(mk_string_tolower(val), allowed_levels[i], val_len)) {
+                                v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
+                                v->via.u64 = (uint64_t)i;
```

Comment:
I think it is not a good idea to change the `msgpack_object` modifying the type , because if we read it later, strange things can happen, it lost the size of the string.

Reply:
I think that is better to use another `msgpack_object` for the level number:

```
diff --git a/src/flb_pack_gelf.c b/src/flb_pack_gelf.c
index 96c1980c..90c9dc05 100644
--- a/src/flb_pack_gelf.c
+++ b/src/flb_pack_gelf.c
@@ -496,6 +496,7 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
 
             msgpack_object *k = &p[i].key;
             msgpack_object *v = &p[i].val;
+            msgpack_object vtmp;
 
             if (k->type != MSGPACK_OBJECT_BIN && k->type != MSGPACK_OBJECT_STR) {
                 continue;
@@ -555,10 +556,27 @@ flb_sds_t flb_msgpack_to_gelf(flb_sds_t *s, msgpack_object *o,
                 else if (v->type == MSGPACK_OBJECT_STR) {
                     val     = v->via.str.ptr;
                     val_len = v->via.str.size;
-                    if (val_len != 1 || val[0] < '0' || val[0] > '7') {
-                        flb_error("[flb_msgpack_to_gelf] level is '%.*s', "
-                                  "but should be in 0..7", val_len, val);
-                        return NULL;
+                    if (!(val_len == 1 && val[0] >= '0' && val[0] <= '7')) {
+                        int n;
+                        char* allowed_levels[] = {
+                            "emerg", "alert", "crit", "err",
+                            "warning", "notice", "info", "debug",
+                            NULL
+                        };
+                        for (n=0; allowed_levels[n] != NULL; n++) {
+                            if (!strncasecmp(val, allowed_levels[n], val_len)) {
+                                v = &vtmp;
+                                v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
+                                v->via.u64 = (uint64_t)n;
+                                break;
+                            }
+                        }
+                        if (allowed_levels[n] == NULL) {
+                            flb_error("[flb_msgpack_to_gelf] level is '%.*s', "
+                                      "but should be in 0..7 or a syslog level",
+                                       val_len, val);
+                            return NULL;
+                        }
                     }
                 }
             }
```

Reply:
I modified my code, but aren't we going to override the old content by rewriting the value of union?
I checked the later assignment of value:
```
else if (v->type == MSGPACK_OBJECT_POSITIVE_INTEGER) {
    val = temp;
    val_len = snprintf(temp, sizeof(temp) - 1,
                        "%" PRIu64, v->via.u64);
}
```
And I don't think this causes any problems.
If you think we can change this back, I think it'd be better to do.

Reply:
@manuelluis Would you please give your feedback about this, too?

Reply:
I don't know if I understood your question, when we do:
```
v = &vtmp;
v->type = MSGPACK_OBJECT_POSITIVE_INTEGER;
v->via.u64 = (uint64_t)n;
```
where are point the `msgpack_object *v` to vtmp that is a struct in the stack, so there  is no allocation in the heap and `v` are not pointing to the `msgpack_object` that we get in the args.

and in:
```
else if (v->type == MSGPACK_OBJECT_POSITIVE_INTEGER) {
    val = temp;
    val_len = snprintf(temp, sizeof(temp) - 1,
                        "%" PRIu64, v->via.u64);
}
```
we are writing over `temp` that is  a char of 48 bytes (enough to write a 64 bit integer) initialized to 0 in the stack, we are not modifying the `msgpack_object`.

Reply:
Yeah, I missed the point that we lose the pointer to that allocated string. You're right.